components/utils.py

```diff
@@ -520,3 +606,282 @@ def untar(source, destination='/tmp', strip=1):
+                                                           self.is_upgrade)
+
+        if not os.path.isfile(local_resource_path):
+            mkdir(os.path.dirname(local_resource_path))
+            if user_resource:
+                self._download_user_resource(source,
+                                             local_resource_path,
+                                             resource_name,
+                                             params)
+            else:
+                self._download_resource(source,
+                                        local_resource_path,
+                                        params)
+            resources_props = self._get_resources_json(service_name,
+                                                       self.is_upgrade)
+            # update the resources.json
+            if resource_name not in resources_props.keys():
+                resources_props[resource_name] = destination
+                self._set_resources_json(resources_props, service_name,
+                                         self.is_upgrade)
+        return local_resource_path, destination
+
+    def _download_user_resource(self, source, dest, resource_name, params):
+        if self.is_upgrade:
+            use_existing = params['use_existing_on_upgrade']
+            if use_existing:
+                service_name = params['service_name']
+                install_props = self.get_install_resources_json(service_name)
+                existing_resource_path = install_props.get(resource_name, '')
+                if os.path.isfile(existing_resource_path):
+                    ctx.logger.info('Using existing resource for {resource}'
+                                    .format(resource=resource_name))
+                    # update the resource file we hold that might have changed
+                    install_resource = self._get_resource_file_path(
+                            service_name, resource_name, False)
+                    copy(existing_resource_path, install_resource)
+                    # copy it to upgrade-resource_name
+                    copy(existing_resource_path, dest)
+                else:
+                    ctx.logger.info('User resource {resource} not found on'
+                                    ' {path}'.format(resource=resource_name,
+                                                     path=dest))
+
+        if not os.path.isfile(dest):
+            self._download_resource(source, dest, params)
+
+    @staticmethod
+    def _download_resource(source, dest, params):
+        resource_name = os.path.basename(dest)
+        ctx.logger.info('Downloading resource {resource} to '
+                        '{dest}'.format(resource=resource_name, dest=dest))
+        ctx.download_resource_and_render(source, dest, params)
+
+    def _get_resource_file_path(self, service_name, resource_name, upgrade):
+        base_service_dir = os.path.join(self.RESOURCES_PATH, service_name)
+        if upgrade:
```

Comment:
In general, I think we shouldn't propagate `update` anywhere and use it as a globally accessible context.
That means that when a script runs, it will know whether it's in the context of an upgrade or not and while importing utils there will already be a context for all of the functions. This will make the code less cumbersome and allow future features in the scripts to just use that context without having to explicitly pass `upgrade` everywhere.